### PR TITLE
Move validation from `pipelinespec` to `pipelinerunspec` when propagating parameters

### PR DESCRIPTION
**Pull request by chitrangpatel**
# Changes

Prior to this, propagating parameters only skipped webhook validation for params defined in script. As a result, when users tried to propagate params to other fields like `args` or `command`, etc. an webhook validation was raised. This PR address issue https://github.com/tektoncd/pipeline/issues/5141. This PR addresses validations in `pipelinespec` and `pipelinerunspec`. The changes for validations in `taskspec` and `taskrunspec` are in a separate [PR](https://github.com/tektoncd/pipeline/pull/5143).

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Move parameter validation from `pipelinespec` to `pipelinerunspec` when propagating parameters
```

### CODE REVIEW

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation.go [hunk 1]
```diff
@@ -107,14 +110,84 @@ func (ps *PipelineRunSpec) Validate(ctx context.Context) (errs *apis.FieldError)
...
 	}
 
 	return errs
 }
 
+// validateInlineParameters validates parameters that are defined inline.
+// This is cruicial for propagated parameters since the parameters could
+// be defined under pipelineRun and then called directly in the task steps.
+// In this case, paratemers cannot be validated by the underlying pipelineSpec
+// or taskSpec since they may not have the parameters declared because of propagation.
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/pipelinerun_validation.go:124]:
```suggestion
// validateInlineParameters validates parameters that are defined inline.
// This is crucial for propagated parameters since the parameters could
// be defined under pipelineRun and then called directly in the task steps.
// In this case, parameters cannot be validated by the underlying pipelineSpec
// or taskSpec since they may not have the parameters declared because of propagation.
```

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go [hunk 2]
```diff
@@ -167,9 +271,15 @@ func TestPipelineRun_Validate(t *testing.T) {
...
+				Params: []v1beta1.Param{{
+					Name: "pipeline-words",
+					Value: v1beta1.ArrayOrString{
+						ArrayVal: []string{"hello", "pipeline"},
+					},
+				}},
 				PipelineSpec: &v1beta1.PipelineSpec{
 					Params: []v1beta1.ParamSpec{{
-						Name: "pipeline-words",
+						Name: "default-pipeline-words",
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:282]: why did you rename this param?

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go [hunk 3]
```diff
@@ -182,20 +292,50 @@ func TestPipelineRun_Validate(t *testing.T) {
 						}},
 						TaskSpec: &v1beta1.EmbeddedTask{TaskSpec: v1beta1.TaskSpec{
 							Params: []v1beta1.ParamSpec{{
-								Name: "task-words",
+								Name: "default-task-words",
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:295]: why did you rename this param?

**Review by jerop — APPROVED** [review]
thanks @chitrangpatel!

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go (hunk 3 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:295]: Same reason as above. Let me know if you need more clarification.

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go (hunk 2 above)
Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:282]: yes, calling it `default` when there's a default field was what was confusing -- yes please rename

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go (hunk 3 above)
Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:295]: the discussion above covered it, thanks

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation.go (hunk 1 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/pipelinerun_validation.go:124]: Thanks!

File: pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go (hunk 2 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go:282]: done

File: pkg/apis/pipeline/v1/pipeline_validation.go [hunk 4]
```diff
@@ -38,6 +39,7 @@ func (p *Pipeline) Validate(ctx context.Context) *apis.FieldError {
 	if apis.IsInDelete(ctx) {
 		return nil
 	}
+	ctx = config.SetValidateParameterVariablesAndWorkspaces(ctx, true)
```
Comment (abayer) [inline pkg/apis/pipeline/v1/pipeline_validation.go:42]: I should have mentioned this on the other PR, but this name is confusing to me. Just based on the name alone, it's not really clear what changes when you set this. It feels like it's saying "ok, validate parameters and workspaces" in general, which seems like something you'd never actually want to turn off. Could we get this a more clear name to better signify exactly what it's doing?

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/pipeline_validation.go:42]: Yes, that's right. How about I have a flag that is defaulted to `true` and we only call it when we intend to make it `false`? 

e.g.

we could have `config.DoNotValidateParameterVariablesAndWorkspaces(ctx)` which would set it the underlying context field to `false`. 

We also have a function `config.ValidateParameterVariablesAndWorkspaces(ctx)` which returns a `true` or `false` based on the value of the context field. This function is called by `task_validation` to decide whether to validate parameters and workspaces. We can have it return `true` by default.

Does that look like a good solution?

Comment (abayer) [reply pkg/apis/pipeline/v1/pipeline_validation.go:42]: I'd really like the name to be more clear than `ValidateParameterVariablesAndWorkspaces` - that name is just too vague. What does it relate to? Which parameter variables and workspaces? If we're just using it for parameter propagation cases, it feels like it would make more sense to have the name be something regarding parameter propagation validation, i.e., _why_ we're setting this flag, not what the result of setting this flag is.

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/pipeline_validation.go:42]: I see what you mean. Its trying to skip validation when it does not contain the full information. e.g. When propagating parameters, a user could provide the params at the task run or the pipeline run level and directly call them in the steps. However, when we validate the underlying taskSpec or pipelineSpec, we don't have these params declared there and so the web hook would throw an error. This flag allows us to specify where to skip validation of params due to lack of information.

I could call the it `SkipValidationDueToPropagatedParametersAndWorkspaces` (a bit of a mouthful, but ok I guess)?

Comment (abayer) [reply pkg/apis/pipeline/v1/pipeline_validation.go:42]: That would definitely be better, yeah.

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/pipeline_validation.go:42]: I will make the appropriate changes.

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1/pipeline_validation.go]: @abayer done! Thanks for the suggestion.

**Comment by JeromeJu** [comment]
It seems that some `ArrayOrString` params in https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/pipelinerun_validation_test.go/#L130 should be updated to `ParamValue`.  

cc @Yongxuanzhang @chitrangpatel Could you please help confirm here?

**Comment by chitrangpatel** [comment]
I agree. I don't recall which PR was merged first. There is backwards compatibility which is why the tests work. However, we should make it ParamValue. I can do that in one shot in a separate PR.

**Comment by JeromeJu** [comment]
Thanks @chitrangpatel
